cvmfs/server/generate_stats_plots.C

```diff
@@ -0,0 +1,578 @@
+    double x, y;
+    gr->GetPoint(i, x, y);
+    int bin = h->FindBin(x);
+    h->SetBinContent(bin, y);
+  }
+  return h;
+}
+
+void WriteHistogram(PlotInstrs pi) {
+  TCanvas *c = new TCanvas(pi.name.c_str());
+  c->SetRightMargin(.02);
+
+  auto hists = new TList();
+
+  int nvals = pi.gr[0]->GetN();
+  double xmin = pi.gr[0]->GetX()[0];
+  // shift xmax to the RIGHT side of the bin
+  double xmax = pi.gr[0]->GetX()[nvals-1] + pi.period;
+  int nbins = (xmax - xmin) / pi.period;
+  auto helper = new TH1D("helper", "", nbins, xmin, xmax);
+  helper->SetBinContent(0, 0);
+  int n_visible_vals = std::min(nbins, max_visible_values);
+
+  auto legend = new TLegend();
+
+  for (unsigned i = 0; i < pi.gr.size(); ++i) {
+    auto gr = *pi.gr[i];
+    auto h = GetHistogram(&gr, pi.period);
+    int nbins = h->GetNbinsX();
+    for (int i = 0; i < nbins; ++i) {
+      if (helper->GetBinContent(i+1) < h->GetBinContent(i+1)) {
+        helper->SetBinContent(i+1, h->GetBinContent(i+1));
+      }
+    }
+
+    h->SetMarkerStyle(pi.marker_styles[i]);
+    h->SetMarkerSize(.7);
+    h->SetMarkerColor(pi.colors[i]);
+    h->SetLineWidth(2);
+    h->SetLineColor(pi.colors[i]);
+    h->SetTitle(pi.gr_titles[i].c_str());
+    hists->Add(h);
+    legend->AddEntry(h, pi.gr_titles[i].c_str());
+  }
+
+  if (pi.time_axis) {
+    helper->GetXaxis()->SetTimeDisplay(1);
+    helper->GetXaxis()->SetTitle("date");
+  } else {
+    helper->GetXaxis()->SetTitle("revision");
+    helper->GetXaxis()->SetNoExponent();
+  }
+  helper->GetXaxis()->CenterTitle();
+  helper->GetXaxis()->SetTitleSize(.05);
+  helper->GetXaxis()->SetRange(nbins - n_visible_vals, nbins);
+
+  helper->GetYaxis()->SetTitle(pi.y_axis_title.c_str());
+  helper->GetYaxis()->SetTitleSize(.05);
+  helper->GetYaxis()->SetTitleOffset(.8);
+  helper->GetYaxis()->CenterTitle();
+  helper->SetMinimum(0);
+
+  helper->SetLabelSize(.045, "xy");
+  helper->SetStats(false);
+  helper->SetLineWidth(0);
+  helper->SetTitle(pi.title.c_str());
+  helper->Draw("L");
+
+  for (int i = 0; i < hists->GetEntries(); ++i) {
+    auto h = hists->At(i);
+    h->Draw("L P0 SAME");
+  }
+
+  if (pi.y_axis_title == "bytes" || pi.y_axis_title == "B/s") {
+    auto *pt = new TPaveText(.8, .12, .96, .32, "N NDC");
+    pt->AddText("10^3 = kB");
+    pt->AddText("10^6 = MB");
+    pt->AddText("10^9 = GB");
+    pt->AddText("10^12 = TB");
+    pt->Draw();
+  }
+
+  if (pi.gr_titles[0] != "") {
+    legend->Draw();
+  }
+  c->Write();
+  for (int i = 0; i < hists->GetEntries(); ++i) {
+    delete hists->At(i);
+  }
+  delete helper;
+}
+
+void WriteTHStack(PlotInstrs pi)
+{
+  TCanvas *c = new TCanvas(pi.name.c_str());
+  c->SetRightMargin(.02);
+
+  double xmin = pi.gr[0]->GetX()[0];
+  double xmax = pi.gr[0]->GetX()[pi.gr[0]->GetN()-1] + pi.period;
+  int nbins = (xmax - xmin) / pi.period;
+  int n_visible_vals = std::min(nbins, max_visible_values);
+  THStack *hs = new THStack("hs", pi.title.c_str());
+
+  auto hists = new TList();
```

Comment:
Is this list needed?

Reply:
yeah, I need to keep the histograms to be able to delete them afterwards. Otherwise, the script prints redefinition warnings.

Reply:
I changed the *TH1D to unique pointer. Still need to hold the ownership till the end of the function, so this is necessary. (or not?)